build.ps1

```diff
@@ -0,0 +1,104 @@
+    [string]$Configuration = 'debug',
+    [switch]$WarnAsError = $true
+)
+
+$NothingToDo = !($Restore -or $Build -or $Test)
+
+# External dependencies
+$sourceNugetExe = "https://dist.nuget.org/win-x86-commandline/v3.3.0/nuget.exe"
+
+# Path variables
+$ProjectRoot = Split-Path -parent $PSCommandPath
+$SolutionFolder = Join-Path $ProjectRoot src
+$SolutionFile = Join-Path $SolutionFolder "PInvoke.sln"
+$ToolsFolder = Join-Path $ProjectRoot tools
+$BinFolder = Join-Path $ProjectRoot "bin"
+$BinConfigFolder = Join-Path $BinFolder $Configuration
+$BinTestsFolder = Join-Path $BinConfigFolder "Tests"
+
+# Set script scope for external tool variables.
+$NuGetPath = Join-Path $ToolsFolder "NuGet.exe"
+$NuGetCommand = Get-Command $NuGetPath -ErrorAction SilentlyContinue
+$MSBuildCommand = Get-Command MSBuild.exe -ErrorAction SilentlyContinue
+$VSTestConsoleCommand = Get-Command vstest.console.exe -ErrorAction SilentlyContinue
+
+Function Get-ExternalTools {
+    if (!$NuGetCommand) {
+        Write-Host "NuGet.exe not found. Downloading to $NuGetPath..."
+        if ($PSCmdlet.ShouldProcess($sourceNuGetExe, "Download")) {
+            Invoke-WebRequest $sourceNugetExe -OutFile $NuGetPath
+        }
+        
+        $NuGetCommand = Get-Command $NuGetPath
+    }
+
+    if (!$MSBuildCommand) {
+        Write-Error "Unable to find MSBuild.exe. Make sure you're running in a VS Developer Prompt."
+        return 1;
+    }
+    
+    if (!$VSTestConsoleCommand) {
+        Write-Error "Unable to find vstest.console.exe. Make sure you're running in a VS Developer prompt."
+        return 2;
+    }
+}
+
+Get-ExternalTools
+
+if ($Restore -and $PSCmdlet.ShouldProcess($SolutionFile, "NuGet restore")) {
+    Write-Output "Restoring NuGet packages..."
+    & $NuGetCommand.Path restore $SolutionFile -Verbosity quiet
+}
+
+if ($Build -and $PSCmdlet.ShouldProcess($SolutionFile, "Build")) {
+    Write-Output "Building..."
+    & $MSBuildCommand.Path $SolutionFile /nologo /nr:false /m /v:minimal /fl "/flp:verbosity=normal;logfile=msbuild.log" "/flp1:warningsonly;logfile=msbuild.wrn" "/flp2:errorsonly;logfile=msbuild.err"
+
+    $warnings = Get-Content msbuild.wrn
+    $errors = Get-Content msbuild.err
+    $WarningsPrompt = "$($warnings.length) warnings during build"
+    $ErrorsPrompt = "$($errors.length) errors during build"
+    if ($errors.length -gt 0) {
+        Write-Error $ErrorsPrompt
+    } else {
+        Write-Output $ErrorsPrompt
+    }
+    
+    if ($WarnAsError -and $warnings.length -gt 0) {
+        Write-Error $WarningsPrompt
+    } elseif ($warnings.length -gt 0) {
+        Write-Warning $WarningsPrompt
+    } else {
+        Write-Output $WarningsPrompt
+    }
+}
+
+if ($Test -and $PSCmdlet.ShouldProcess('Test assemblies', 'vstest.console.exe')) {
+    $TestAssemblies = Get-ChildItem -Recurse "$BinTestsFolder\*.Tests.dll"
+    Write-Output "Testing..."
+    & $VSTestConsoleCommand.Path /Parallel /TestAdapterPath:$BinTestsFolder $TestAssemblies
```

Comment:
On my system vstest complains with :

```
Microsoft (R) Test Execution Command Line Tool Version 14.0.24720.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Error: Unrecognized parameter "/Parallel".
```

But I don't have the 2015 Update 2 preview installed so maybe it's due to that ?